db/db_iter.cc

```diff
@@ -470,11 +483,19 @@ inline bool DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check)
     }
 
     if (IsVisible(ikey_.sequence)) {
-      if (skipping && user_comparator_.Compare(ikey_.user_key,
-                                               saved_key_.GetUserKey()) <= 0) {
+      // If the previous entry is of seqnum 0, the current entry will not
+      // possibly be skipped. This condition can potentially be relaxed to
+      // prev_key.seq <= ikey_.sequence. We are cautious because it will be more
+      // prone to bugs causing the same user key with the same sequence number.
+      if (!is_prev_key_seqnum_zero && skipping &&
+          user_comparator_.Compare(ikey_.user_key, saved_key_.GetUserKey()) <=
+              0) {
         num_skipped++;  // skip this entry
         PERF_COUNTER_ADD(internal_key_skipped_count, 1);
       } else {
+        assert(!is_prev_key_seqnum_zero ||
```

Comment:
If is_prev_key_seqnum_zero is an optimization, should not ::Compare return > 0 here whether is_prev_key_seqnum_zero is true or not?

Reply:
It's a good point. I can improve it. What makes it complicated is `skipping`. I am not sure we always have `saved_key_` available here. But in the case of `is_prev_key_seqnum_zero`, we do have. Maybe I can add a `!skipping` check here to be safe.